### DISCUSSION TITLE
Request to have a copy feedback message using the same elements from a tooltip

### DISCUSSION THREAD

**Discussion by telosantos** [iTwin/iTwinUI#2059] [Ideas]
### Feature

Mathieu Fournier is implementing a copy/paste functionality, and we have decided to use a feedback message when objects are copied to signify, inform users they successfully copied them. It's a minor change by reusing the same css properties of the tooltip component, with the addition of a check icon.

**Comment by Ben-Pusey-Bentley** [comment]
Hello, thanks for opening an issue with us. I think that what you are looking for is actually closer to our [toast](https://itwinui.bentley.com/docs/toast) component than a tooltip. Is there a reason that the toast component does not work for this specific use case?

**Reply by telosantos** [reply]
yes, however, the notification is located on the bottom right, which is not helpful for quick interactions like copying elements. would it be customizable to position the feedback message above the cursor or on top of the copied elements? @mathieu-fournier

**Reply by Ben-Pusey-Bentley** [reply]
There are position customization options available for the toast component, but they are all currently relative to the edges of the container (i.e. top, bottom left). We could investigate adding the option to position the message based on the cursor. @mayank99 Do you think this makes sense to add as an option?

**Reply by mayank99** [reply]
@telosantos Could you share more details about this scenario? What are the desired semantics of these controls and how would you operate it without a mouse? A video recording would also help us understand what exactly is being copied and how it's triggered.

Generally speaking, I'm leaning towards considering this out-of-scope / anti-pattern from the perspective of iTwinUI, for multiple reasons:
- Tooltips are for labels and descriptions on _interactive_ elements (like buttons). We have published [detailed guidance](https://itwinui.bentley.com/docs/tooltip#accessibility) on this topic.
- We currently do not provide any components for use on canvas elements. This kind of thing might fall under [AppUI](https://github.com/iTwin/appui) territory. When you force iTwinUI components to be (mis)used against how they were designed to be used, it will usually lead to all sorts of other issues.

**Reply by mathieu-fournier** [reply]
We are copying stuff to the clipboard with `ctrl-c` and came to the conclusion that this was a good way to give feedback tot he user that something happened with the shortcut. 

It's a feedback for a shortcut that would be otherwise be completely silent and feedbackless.

Is this an anti-pattern ? 
I'll let you decide if it's valid or not, you're the experts here hehe. (:

**Reply by mayank99** [reply]
I don't think using a keyboard-shortcut should trigger a tooltip anchored to an element. A toast notification is the better, more predictable choice for this. What you should try to ensure is that the currently selected is very prominently highlighted, so that the user knows what is getting copied when they see the toast.

But just to confirm, what is the purpose of copying it? Can they paste it back into the canvas or is it for pasting into a different context? And do you provide another way of copy/paste-ing (e.g. right click menu, toolbar buttons)?

To me, copy-paste feels like such a common and frequent operation that it shouldn't require feedback. It would actually be distracting to see this feedback repeatedly if you're moving things around on the canvas a lot. But if it's an infrequent and more impactful operation (e.g. copying a large set of elements for pasting into a spreadsheet, almost like an "Export"), then it makes sense to show a toast notification.

**Reply by mathieu-fournier** [reply]
It's to do both. They can paste it right away of whenever they what in another app from the clipboard.

No other way to copy-paste. 

No-feedback might be the best here. I feel like a toast would be too intrusive/distracting for such a simple and usual feature.

**Reply by mayank99** [reply]
> No other way to copy-paste. 

You should provide another way to copy-paste. A keyboard shortcut should never be the only way to perform an action, as it will exclude anybody who isn't using a keyboard or isn't aware of the shortcuts.

> I feel like a toast would be too intrusive/distracting for such a simple and usual feature.

You could show the toast only when more than N items are copied/pasted.

**Comment by r100-stack** [comment]
Slightly related: https://github.com/iTwin/iTwinUI/issues/1055, https://github.com/iTwin/iTwinUI/issues/1055#issuecomment-1430439521
